### ISSUE TITLE
persistent undo

### ISSUE THREAD

**Issue by trusktr** [asvetliakov/vscode-neovim#474]
Would be great to have this in VS Code.

I enjoy in plain (neo)vim when I've set my undo path and have and

```viml
      if has('persistent_undo')
        set undofile	" keep an undo file (undo changes after closing)
      endif
```

so that every time I re-open a file any time in the future (even after closing vim/neovim) the undo history persists.

I have that setting in my `init.vim`, but looks like it doesn't work in VS Code Neo Vim.

**Comment by asvetliakov** [comment]
From nvim documentation looks like it saves undo history only when writing a buffer to a file. The problem neovim never writes anything by design

**Comment by trusktr** [comment]
Ooh, I see, and it then maps the written file to the respective file in the undo dir. I suppose a solution could be, although it adds disk I/O, would be to write the files somewhere unique to Neo Vim extension, basically mapping from the actual file's path to a Neo Vim specific path, and that would implicitly make the native undo history persist when neovim writes to the third location.

**Comment by asvetliakov** [comment]
> suppose a solution could be, although it adds disk I/O, would be to write the files somewhere unique to Neo Vim extension, basically mapping from the actual file's path to a Neo Vim specific path, and that would implicitly make the native undo history persist when neovim writes to the third location.

That is possible workaround, yes. But it's totally unacceptable, otherwise the extension will become like malware storing possibly sensitive information in it's own place.

**Comment by trusktr** [comment]
f.e. `~/path/to/original/file.js` -> `~/.vscode-neovim/path/to/original/file.js` -> `~/path/to/.vim/undo/path%to%original%file.js`

(vim/neovim replaces the slashes to make it a single file).

**Comment by trusktr** [comment]
I think it would be acceptable if it was an option with clear explanation so users know what to expect.

Another idea is, what if neovim just writes to the same file a second time after VS Code has written to it?

**Comment by asvetliakov** [comment]
That also wouldn't work in all cases, because vscode supports with wider file protocols than neovim, like remote or wsl. Neovim can't write to them

**Comment by asvetliakov** [comment]
ideally neovim must support providing own file layer implementation, like `vim.api.fileSave = ....` . Which will allow to make much much better interop. But don't expect it would be added in near time, if ever.

**Comment by trusktr** [comment]
> like remote or wsl

Can those types be detected and the behavior just not applied in that case?

I personally don't mind the double-writing if it means I can gain that feature. I know it isn't an ideal way to implement it, but the result would be totally useful.

**Comment by asvetliakov** [comment]
>Can those types be detected and the behavior just not applied in that case?

Possible. Anyway that would bring the problem *who is the owner and is responsible for that file?* . I don't think this is a viable choice. Will be better to just ask neovim devs to add/expand undo api.

**Comment by trusktr** [comment]
> who is the owner and is responsible for that file?

What do you mean by that? I imagine, like with vim/neovim, the full absolute file path is encoded in the user's home folder (with the `%` for the path separators). Then the history is unique only to the user who edits the file. vim/neovim will warn if the file is opened and doesn't match the undo state (in case some other person modified a root file too).

If I understand what you mean, then vasically vim/neovim already handles that who-owns question.

> Will be better to just ask neovim devs to add/expand undo api.

Perhaps VS Code devs too, so we can make newly-opened files have pre-determined edit histories.

**Comment by David-Else** [comment]
VS Code recently added persistent undo, your wish should have already been granted!

I turn it off as it really annoys me, when I close I file I want that to be the final undo: `"files.restoreUndoStack": false`. I thought it was `true` by default, check it out.

**Comment by theol0403** [comment]
The problem is that the vscode undo stack isn't loaded back into the neovim
buffer, so unless you use the vscode undo functionality, it won't work.

**Comment by trusktr** [comment]
> VS Code recently added persistent undo, your wish should have already been granted!

True, but it is only for when re-opening a file in the same VS Code session without closing VS Code. Once VS Code is closed, then history is lost, and next startup of VS Code has no history. They will add that feature in 

https://github.com/microsoft/vscode/issues/7536.

**Comment by trusktr** [comment]
> The problem is that the vscode undo stack isn't loaded back into the neovim buffer, so unless you use the vscode undo functionality, it won't work.

Question: can we arbitrarily replace neovim's history? (basically the other way around). If so, then that may be a way to rely on VS Code's semi-persistent undo (and eventually its fully-persistent undo), although in this case we would lose the history tree. But it could still be a worthy option for those that are okay with not having the tree.

**Comment by gaoqiangks** [comment]
I add the following into my init.vim, and it seems to be a  workaround for those who really need it.  However, I'm not quite familar with vimscript. Hope for someone to give a usable solution.

`function VscodeUndoRead()
	let g:buf=expand("%")
	let g:ufile1=&undodir."/".g:buf
	let g:ufile=&undodir."/". substitute(g:buf,'[^a-zA-Z]', '_', 'g')
	if filereadable(g:ufile)
		call execute("rundo ".g:ufile)
	endif
	"echo "VscodeUndoRead"
endfunction
function VscodeUndoWrite()
	let g:buf=expand("%")
	let g:ufile1=&undodir."/".g:buf
	let g:ufile=&undodir."/". substitute(g:buf,'[^a-zA-Z]', '_', 'g')
	call execute("wundo ".g:ufile)
	"echo "VscodeUndoWrite"
endfunction
if g:in_vscode==1
	"I don't which event I should use since some events in vscode-neovm don't work in the
	"way I'm familar with.
	autocmd BufEnter * call VscodeUndoRead()
	autocmd BufUnload * call VscodeUndoWrite()
	autocmd InsertLeave * call VscodeUndoWrite() 
endif`

**Comment by justinmk** [comment]
For this work, Nvim would need to enhance the 'undofile' option to save undo info [for buftype=acwrite files](https://github.com/vscode-neovim/vscode-neovim/issues/1260#issuecomment-1638381606).

Can you open a feature request at https://github.com/neovim/neovim

**Comment by justinmk** [comment]
> Next, we need to install the [multi-command](https://marketplace.visualstudio.com/items?itemName=ryuta46.multi-command) extension and define a multi-command in user settings that calls the `VSCodeWriteUndoFile` every time the multi-command is run.

vscode has `runCommands` https://github.com/microsoft/vscode/pull/177104 now I think? Though I can't find docs on it.

@MahdiNazemi your solution seems pretty reasonable , create a PR?

**Comment by justinmk** [comment]
>  Do you have any pointers to which files I should start looking at?

it's a very small codebase (intentionally), just look around :)

>  How does vscode-neovim deal with `:w`?

it defines a `BufWriteCmd` autocmd/event handler https://github.com/vscode-neovim/vscode-neovim/pull/2116



>  Do you intend to keep `clear_undo` and call it [here](https://github.com/vscode-neovim/vscode-neovim/blob/bb0389ac13c5215280cd0f66b32e80e3c916055c/src/buffer_manager.ts#L779-L781)?

maybe @xiyaowong can comment

**Comment by theol0403** [comment]
@MahdiNazemi thanks for looking into this!
* this should be implemented using the lua API. will remove the need for ES
* runCommands should not be used. you can implement the same but with the sequencing happening inside lua. however, the better approach is to add to the `BufWriteCmd` hook
* `clear_undo` is used because when initializing the buffer, vscode-neovim copies everything from vscode to neovim. if it wasn't called, you would be able to undo all the way to an empty file
